src/algebra/module/torsion.lean

```diff
@@ -117,12 +146,91 @@ lemma torsion_by_is_torsion_by : is_torsion_by R (torsion_by R M a) a := λ _, s
 
-end torsion_by
+lemma torsion_by_le_torsion_by_of_dvd (a b : R) (dvd : a ∣ b) :
+  torsion_by R M a ≤ torsion_by R M b :=
+λ x hx, begin
+  cases dvd with c h,
+  change _ • _ = _, change _ • _ = _ at hx,
+  rw [h, mul_comm, mul_smul, hx, smul_zero]
+end
+
+lemma torsion_by_one : torsion_by R M 1 = ⊥ :=
+eq_bot_iff.mpr (λ _ h, by { rw [mem_torsion_by_iff, one_smul] at h, exact h })
+
+section coprime
+open_locale big_operators
+open dfinsupp
+variables {ι : Type*} {p : ι → R} {S : finset ι} (hp : pairwise (is_coprime on λ s : S, p s))
+include hp
 
-section quotient
+lemma supr_torsion_by_eq_torsion_by_prod :
+  (⨆ i : S, torsion_by R M (p i)) = torsion_by R M (∏ i in S, p i) :=
+begin
+  cases S.eq_empty_or_nonempty with h h,
+  { rw [h, finset.prod_empty, torsion_by_one],
+    convert supr_of_empty _, exact subtype.is_empty_false },
+  apply le_antisymm,
+  { apply supr_le _, rintro ⟨i, is⟩,
+    exact torsion_by_le_torsion_by_of_dvd _ _ (finset.dvd_prod_of_mem p is) },
+  { intros x hx, classical, rw mem_supr_iff_exists_dfinsupp',
+    cases (exists_sum_eq_one_iff_pairwise_coprime h).mpr hp with f hf,
+    use equiv_fun_on_fintype.inv_fun (λ i, ⟨(f i * ∏ j in S \ {i}, p j) • x, begin
+      obtain ⟨i, is⟩ := i,
+      change p i • (f i * ∏ j in S \ {i}, _) • _ = _, change _ • _ = _ at hx,
+      rw [smul_smul, mul_comm, mul_assoc, mul_smul, ← finset.prod_eq_prod_diff_singleton_mul is,
+        hx, smul_zero]
+    end⟩),
+    simp only [equiv.inv_fun_as_coe, sum_eq_sum_fintype, coe_eq_zero, eq_self_iff_true,
+      implies_true_iff, finset.univ_eq_attach, equiv_fun_on_fintype_apply],
+    change ∑ i : S, ((f i * ∏ j in S \ {i}, p j) • x) = x,
+    have : ∑ i : S, _ = _ := S.sum_finset_coe (λ i, f i * ∏ j in S \ {i}, p j),
+    rw [← finset.sum_smul, this, hf, one_smul] }
+end
 
+lemma torsion_by_independent : complete_lattice.independent (λ i : S, torsion_by R M (p i)) :=
+λ i, begin
+  classical,
+  dsimp, rw [disjoint_iff, eq_bot_iff], intros x hx,
+  rw submodule.mem_inf at hx, obtain ⟨hxi, hxj⟩ := hx,
+  have hxi : p i • x = 0 := hxi,
+  rw mem_supr_iff_exists_dfinsupp' at hxj, cases hxj with f hf,
+  obtain ⟨b, c, h1⟩ := pairwise_coprime_iff_coprime_prod.mp hp i i.2,
+  rw [mem_bot, ← one_smul _ x, ← h1, add_smul],
+  convert (zero_add (0:M)),
+  { rw [mul_smul, hxi, smul_zero] },
+  { rw [← hf, smul_sum, sum_eq_zero],
+    intro j, by_cases ji : j = i,
+    { convert smul_zero _,
+      rw ← mem_bot _, convert coe_mem (f j),
+      symmetry, rw supr_eq_bot, intro hj',
+      exfalso, exact hj' ji },
+    { have hj' : ↑j ∈ S \ {i},
+      { rw finset.mem_sdiff, refine ⟨j.2, λ hj', ji _⟩, ext, rw ← finset.mem_singleton, exact hj' },
+      rw [finset.prod_eq_prod_diff_singleton_mul hj', ← mul_assoc, mul_smul],
+      have : (⨆ (H : ¬j = i), torsion_by R M (p j)) ≤ torsion_by R M (p j) := supr_const_le,
+      have : (f j : M) ∈ torsion_by R M (p j) := this (coe_mem _),
+      have : _ • _ = _ := this,
```

Comment:
Does this work?
```suggestion
      have : _ • _ = _ := this (coe_mem _),
```

Reply:
I remember it didn't work when I was doing it but now it works, weird